### PR TITLE
fix(task): require trust for remote metadata

### PR DESCRIPTION
**Pull request by risu729**
Replaces the auto-closed #11113.

## Work order

1. #12254 — contain remote Git task paths (independent)
2. This PR — require trust before remote metadata discovery
3. #12256 — guard remote metadata side effects
4. #12257 — resolve trusted metadata in task commands
5. #12206 — harden cache lifecycle after the trust stack is merged and rebased

#12254 is independent and may merge separately. #12256 and #12257 are stacked on this PR; GitHub shows cumulative diffs against main because the branches live in a fork.

## Summary

- require trust in the defining config before passive commands fetch remote HTTP or Git tasks (`mise tasks ls` / `info` / `deps` / `validate`, plus `mise cache task` / `mise cache clear <task>`)
- require trust before decoded remote header templates render, and keep that gate closed in safe mode
- preserve defining-config provenance through explicitly included and nested remote task files
- keep the fetched task artifact associated with that provenance in cached and no-cache modes
- fail fast in `mise watch` under safe mode instead of expanding dependencies before the spawned `mise run` refuses

## Safe mode

Safe mode is a code-execution boundary, not a network boundary. An untrusted config can already resolve and download arbitrary URLs there through the HTTP-based backends — `mise install` writes an http-backend tool's artifact to disk with no trust, and `docs/security.md` documents that as intended so `mise lock` can run against pull-request config. Gating the remote task fetch specifically would have contradicted a capability the same untrusted config already has in the same mode, and it broke passive introspection wholesale: a single remote task made `mise tasks ls` fail for every task, including local ones.

So the split is:

- **fetch** — `remote_fetch_allowed`: safe mode waives it, otherwise the defining config must already be trusted. This *authorizes* rather than *acquires*: passive discovery deliberately does not use `trust_check`, because its prompt (or `MISE_YES=1`) would persist trust for the whole config root and turn a read-only `mise tasks ls` into a permanent grant of env, hooks and templates. Passive commands are absent from `implicitly_trusts_active_config` for the same reason.
- **act** — `trust_check_remote_content`, which safe mode does *not* waive, applied when `remote_header_requires_trust` says the fetched header can do something before anyone runs the task. Two channels can: a decoded `#MISE` template (renders at discovery — code execution, file reads, process env interpolated into printed metadata) and a `#USAGE include`, which usage-lib resolves against an arbitrary absolute path while the spec is parsed for display. Unlike a local template, the payload is invisible in the config that names it and can vary per request.

### Trust in CI

These gates are trust-based, so in CI they are inert: `is_trusted` blanket-trusts any checkout under `CI=true`, which is how mise works in CI generally. `MISE_PARANOID=1` is the existing mechanism if you want trust enforced there — its branch runs before the CI shortcut, so the gates do hold under paranoid (verified).

Worth being explicit that safe mode alone does not stop metadata from interpolating the process env into printed output: `exec()` and `read_file()` are stubbed, but `{{ env.X }}` / `get_env()` are not, so a description can carry a secret into CI logs. That is a pre-existing property of safe mode and applies identically to a plain local config, so it is not addressed here.

This mirrors `task_include_requires_trust`, which already treats template-free task files as inert and asks for trust only when a file contains templates.

Cloning a `git::` include stays gated in safe mode. Passing that gate marks the includes authorized, which clears `require_task_include_trust` for the whole include set, so tasks from the clone render with no further check — the clone gate is all there is. Relaxing it means first giving included tasks their own safe-mode-aware render gate *and* decoupling that from `remote_includes_authorized`, so it is left as follow-up.

The `mise watch` change is a fail-fast/UX fix rather than a security fix — watch is useless in safe mode either way, since the `mise run` it spawns refuses; erroring up front just avoids the pointless dependency expansion first. (Its commit message overstates this; treat this section as the accurate framing.)

## Root cause

#11111 began parsing remote task headers during discovery. Passive discovery could then perform HTTP or Git work before the local config selecting that task had been trusted, and nested remote task files could lose the original trust authority.

## Split boundaries

- remote Git path containment is isolated in #12254
- side-effectful env and tool consumers are isolated in #12256
- aliases, visibility, command dispatch, dependency fallback, and snapshot-sensitive discovery are isolated in #12257
- cache publication, locking, retention, and cache-key hardening remain in #12206

## Validation

- mise run lint-fix
- mise run test:unit task::task_fetcher::tests
- mise run test:unit test_remote_fetch_trust_policy
- mise run test:e2e e2e/tasks/test_task_remote_metadata_trust e2e/tasks/test_task_remote_git_includes

Follow-up to #11111.

*AI-assisted — Tool: Codex; model: unavailable/unavailable; version: unavailable.*
*AI-assisted — Tool: Claude Code; model: anthropic/claude-fable-5; version: 2.1.239.*

## Summary by CodeRabbit

* **New Features**
  * Added stronger trust controls for remote task files, metadata, templates, and includes.
  * Added support for nested remote task includes and improved dependency displays.
  * Remote tasks now retain source information for trust decisions.
  * Safe mode can list inert remote tasks without granting trust.

* **Bug Fixes**
  * Prevented untrusted remote content from being rendered or executed.
  * Updated listing, validation, information, caching, dependency, and watch commands to honor trust requirements.
  * Improved cleanup, request handling, and snapshot behavior after remote task failures.

### CODE REVIEW

**Comment by risu729** [comment]
@coderabbitai resume

File: e2e/helpers/scripts/git_http_backend_server.py [hunk 1]
```diff
@@ -203,18 +220,17 @@ def start_server(port=0):
     repo_path = temp_dir / 'repo'
     upload_pack_count_path = os.environ.get('MISE_GIT_HTTP_UPLOAD_PACK_COUNT_FILE')
     upload_pack_count_file = Path(upload_pack_count_path) if upload_pack_count_path else None
+    request_log = os.environ.get('MISE_GIT_HTTP_REQUEST_LOG')
     if upload_pack_count_file is not None:
         upload_pack_count_file.parent.mkdir(parents=True, exist_ok=True)
         upload_pack_count_file.write_text('')
```
[the remark this answers is not in the document]
Comment (risu729) [reply e2e/helpers/scripts/git_http_backend_server.py]: Fixed in b96801dd1 — the request log now gets `parent.mkdir(parents=True, exist_ok=True)` + `write_text("")` at startup, matching the `upload_pack_count_file` handling directly above it, and it is unlinked in the same `finally` block for symmetry. It also closes a hole in the test itself: `test -s` on a missing file was indistinguishable from an empty one, so the "no request was made" assertion could have passed because the server never created the log.

*AI-assisted — Tool: Claude Code; model: anthropic/claude-fable-5; version: 2.1.239.*

File: e2e/helpers/scripts/git_http_backend_server.py [hunk 2]
```diff
@@ -223,6 +239,8 @@ def handler(*args, **kwargs):
     # This avoids race conditions between finding and binding
     with socketserver.TCPServer(("", port), handler) as httpd:
         actual_port = httpd.server_address[1]
+        print(f"Creating test repository at {repo_path}")
+        create_test_repo(str(repo_path), actual_port)
```
[the remark this answers is not in the document]
Comment (risu729) [reply e2e/helpers/scripts/git_http_backend_server.py:247]: Not addressed here: `create_test_repo` already ran outside the `try`/`finally` before this PR (it was called just above the `with socketserver.TCPServer(...)` block), so the leak on a startup failure is pre-existing and not something this change introduces. This PR only moved the call inside the `with` block so the repository can embed the OS-assigned port. Leaving it out to keep the diff scoped; happy to fix it separately if you want.

*AI-assisted — Tool: Claude Code; model: anthropic/claude-fable-5; version: 2.1.239.*
